src/object.d

```diff
@@ -3241,6 +3241,226 @@ template RTInfo(T)
+    static if ((is(U1 == float) && is(U2 == float))
+        || (is(U1 == double)   && is(U2 == double))
+        || (is(U1 == real)     && is(U2 == real))
+        || (is(U1 == ifloat)   && is(U2 == ifloat))
+        || (is(U1 == idouble)  && is(U2 == idouble))
+        || (is(U1 == ireal)    && is(U2 == ireal))
+        || (is(U1 == cfloat)   && is(U2 == cfloat))
+        || (is(U1 == cdouble)  && is(U2 == cdouble))
+        || (is(U1 == creal)    && is(U2 == creal)))
+    {
+        template Floating(T)
+        if (is(T == cfloat) || is(T == cdouble) || is(T == creal))
+        {
+          pure nothrow @safe:
+            int compare(T f1, T f2)
+            {
+                int result;
+
+                if (f1.re < f2.re)
+                    result = -1;
+                else if (f1.re > f2.re)
+                    result = 1;
+                else if (f1.im < f2.im)
+                    result = -1;
+                else if (f1.im > f2.im)
+                    result = 1;
+                else
+                    result = 0;
+                return result;
+            }
+        }
+
+        template Floating(T)
+        if (is(T == float) || is(T == double) || is(T == real))
+        {
+          pure nothrow @safe:
+            int compare(T d1, T d2)
+            {
+                if (d1 != d1 || d2 != d2) // if either are NaN
+                {
+                    if (d1 != d1)
+                    {
+                        if (d2 != d2)
+                            return 0;
+                        return -1;
+                    }
+                    return 1;
+                }
+                return (d1 == d2) ? 0 : ((d1 < d2) ? -1 : 1);
+            }
+        }
+
+        template Array(T)
+        if (is(T ==  float) || is(T ==  double) || is(T ==  real) ||
+            is(T == cfloat) || is(T == cdouble) || is(T == creal))
+        {
+          pure nothrow @safe:
+            int compare(T[] s1, T[] s2)
+            {
+                size_t len = s1.length;
+                if (s2.length < len)
+                    len = s2.length;
+                for (size_t u = 0; u < len; u++)
+                {
+                    if (int c = Floating!T.compare(s1[u], s2[u]))
+                        return c;
+                }
+                if (s1.length < s2.length)
+                    return -1;
+                else if (s1.length > s2.length)
+                    return 1;
+                return 0;
+            }
+        }
+
+        static if (is(U1 == ifloat)) alias F = float;
+        else static if (is(U1 == idouble)) alias F = double;
+        else static if (is(U1 == ireal)) alias F = real;
+        else alias F = U1;
+
+        return () @trusted { return Array!F.compare(cast(F[])s1, cast(F[])s2); }();
+    }
+
+    // integral, char types
+    else
+    {
+        size_t len = s1.length;
+
+        if (s2.length < len)
+            len = s2.length;
+
+        for (size_t u = 0; u < len; u++)
```

Comment:
I agree with jacob. We should not add more and more stuff to `object.d` and pollute the global namespace if we can easily avoid it. Lowering does not require the symbol to be visible to user written D code. We already do something similar in GDC to load the shared library types symbol from `rt.sections`, no need to place it into `object.d`:
```c++
  // Try to find the required types and functions in druntime
  Dsymbol *mref = NULL, *compiler_dso_type = NULL, *dso_registry_func = NULL;
  // Ignore error if the file can not be found and simply don't emit Moduleinfo
  unsigned errors = global.startGagging();
  Module *m = Module::load(Loc(), NULL, Id::sectionsModule); //Just load the module in the compiler frontend
  global.endGagging(errors);//

  if (m)
    {
      m->importedFrom = m;
      m->importAll (NULL);
      m->semantic(NULL);
      m->semantic2(NULL);
      m->semantic3(NULL);
      // These symbols are normally private to the module they're declared in,
      // but for internal compiler lookups, their visibility is discarded.
      int sflags = IgnoreErrors | IgnoreSymbolVisibility;
      mref = m->search(Loc(), Id::Dmodule_ref, sflags);
      compiler_dso_type = m->search(Loc(), Id::compiler_dso_type, sflags);
      dso_registry_func = m->search(Loc(), Id::dso_registry_func, sflags);
    }
```

This can certainly work for templates to in some case. From a high level POV this just means looking up these special symbols does not require an explicit import.